codalab/worker/local_run/docker_image_manager.py

```diff
@@ -102,13 +104,20 @@ def _cleanup(self):
                     entry_to_remove.digest,
                 )
                 try:
-                    image_to_delete = self._docker.images.get(entry_to_remove.id)
-                    tags_to_delete = image_to_delete.tags
-                    for tag in tags_to_delete:
-                        self._docker.images.remove(tag)
-                    # if we successfully removed the image also remove its cache entry
+                    # We try to delete images from image cache in LRU order
+                    self._docker.images.remove(entry_to_remove.id)
                     del self._image_cache[entry_to_remove.digest]
-                except docker.errors.NotFound:
+
+                    # We try to remove any dangling docker images
+                    deleted_dangling_entries = self._docker.images.prune(filters={'dangling': True})
+                    if deleted_dangling_entries['ImagesDeleted'] is not None:
+                        for deleted_entry in deleted_dangling_entries['ImagesDeleted']:
+                            if deleted_entry['Deleted'] in self._image_cache:
+                                # if we successfully removed the dangling image which exists in cache entry,
+                                # we should also remove its cache entry
+                                del self._image_cache[deleted_entry['Deleted']]
+                                deletable_entries.remove(deleted_entry['Deleted'])
+                except docker.errors.ImageNotFound:
                     # image doesn't exist anymore for some reason, stop tracking it
                     del self._image_cache[entry_to_remove.digest]
                 except docker.errors.APIError as err:
```

Comment:
Also, forcing removing a running docker image won't give us any benefit, as it only removes the tag of the running image, not the image itself.